### PR TITLE
gh-125916: Allow functools.reduce 'initial' to be a keyword argument

### LINKED ISSUE #125916: Allow `functools.reduce`s 'initial' to be a keyword argument
# Feature or enhancement

### Proposal:

`functools.reduce` takes `function` (generally a callable) and `iterable`, with an optional `initial` parameter:

https://docs.python.org/3/library/functools.html#functools.reduce

However, `initial` cannot be passed as a keyword argument, which reduces ;) readability,
```python
from functools import reduce
from operator import sub
>>> reduce(sub, [1, 1, 2, 3, 5, 8], 21)
1
>>> reduce(sub, [1, 1, 2, 3, 5, 8], initial=21)
TypeError: reduce() takes no keyword arguments
```

Allowing `initial` as keyword argument will be more clear, and `initial` could also be passed as a keyword argument while making `partial` functions out of `reduce`.

### Has this already been discussed elsewhere?

I have already discussed this feature proposal on Discourse

### Links to previous discussion of this feature:

https://discuss.python.org/t/remove-positional-only-restriction-on-initial-parameter-of-reduce/68897

### Linked PRs
* gh-125917
* gh-125999

### PR DESCRIPTION
**Pull request by sayandipdutta**
## Before:
```python
from functools import reduce
from operator import sub
>>> reduce(sub, [1, 1, 2, 3, 5, 8], 21)
1
>>> reduce(sub, [1, 1, 2, 3, 5, 8], initial=21)
TypeError: reduce() takes no keyword arguments
```

## After:
```python
from functools import reduce
from operator import sub
>>> reduce(sub, [1, 1, 2, 3, 5, 8], 21)
1
>>> reduce(sub, [1, 1, 2, 3, 5, 8], initial=21)
1
```

Issue: gh-125916

----
📚 Documentation preview 📚: https://cpython-previews--125917.org.readthedocs.build/

### CODE REVIEW

File: Modules/_functoolsmodule.c
Comment (skirpichev) [inline Modules/_functoolsmodule.c]:
```suggestion
    {"reduce", _PyCFunction_CAST(functools_reduce), METH_VARARGS|METH_KEYWORDS,
```

File: Doc/library/functools.rst
Comment (skirpichev) [inline Doc/library/functools.rst]:
```suggestion

   .. versionchanged:: 3.14
      *initial* is now supported as a keyword argument.
```

I think this should fix docs error.

File: Modules/_functoolsmodule.c
Comment (skirpichev) [inline Modules/_functoolsmodule.c]:
```suggestion
    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "OO|O:reduce", keywords,
                                     &func, &seq, &result))
    {
```
PEP 7 nitpick.

**Review by skirpichev — COMMENTED** [review]
Please also benchmark your implementation.

Kwargs handling *will* affect performance even if keyword will not be actually used (e.g. calls like ``reduce(f, seq, init)``).  IIUC, ``PyArg_ParseTupleAndKeywords`` is much slower in general than ``PyArg_UnpackTuple``.

**Comment by skirpichev** [comment]
CC @sobolevn, as you have added AC comment.

**Comment by skirpichev** [comment]
I did some benchmarks.

<details>

```py
# a.py
import pyperf
from functools import reduce

f = lambda x, y: x + y
lst = list(range(10))
init = 123

runner = pyperf.Runner()
runner.bench_func('reduce(f, lst)', reduce, f, lst)
runner.bench_func('reduce(f, lst, init)', reduce, f, lst, init)
```

Run e.g. with: ``python a.py -q -o ref.json``.

</details>

with results:
| Benchmark            | ref     | patch                 |
|----------------------|:-------:|:---------------------:|
| reduce(f, lst)       | 2.18 us | 2.42 us: 1.11x slower |
| reduce(f, lst, init) | 2.35 us | 2.64 us: 1.12x slower |
| Geometric mean       | (ref)   | 1.12x slower          |

**Comment by skirpichev** [comment]
Now with AC (patch2).

<details>

<summary>Patch with AC (a draft).</summary>

```diff
diff --git a/Modules/_functoolsmodule.c b/Modules/_functoolsmodule.c
index 802b1cf792..8faa8ad1ac 100644
--- a/Modules/_functoolsmodule.c
+++ b/Modules/_functoolsmodule.c
@@ -932,15 +932,30 @@ _functools_cmp_to_key_impl(PyObject *module, PyObject *mycmp)
 
 /* reduce (used to be a builtin) ********************************************/
 
-// Not converted to argument clinic, because of `args` in-place modification.
-// AC will affect performance.
+/*[clinic input]
+_functools.reduce
+
+    function as func: object
+    iterable as seq: object
+    /
+    initial as result: object(c_default="NULL") = None
+
+Apply a function of two arguments cumulatively.
+
+Apply it to the items of a sequence or iterable, from left to right, so as to
+reduce the iterable to a single value.  For example, reduce(lambda x, y: x+y,
+[1, 2, 3, 4, 5]) calculates ((((1+2)+3)+4)+5).  If initial is present, it is
+placed before the items of the iterable in the calculation, and serves as a
+default when the iterable is empty.
+[clinic start generated code]*/
+
 static PyObject *
-functools_reduce(PyObject *self, PyObject *args)
+_functools_reduce_impl(PyObject *module, PyObject *func, PyObject *seq,
+                       PyObject *result)
+/*[clinic end generated code: output=30d898fe1267c79d input=b7082b8b1473fdc2]*/
 {
-    PyObject *seq, *func, *result = NULL, *it;
+    PyObject *args, *it;
 
-    if (!PyArg_UnpackTuple(args, "reduce", 2, 3, &func, &seq, &result))
-        return NULL;
     if (result != NULL)
         Py_INCREF(result);
 
@@ -1006,16 +1021,6 @@ functools_reduce(PyObject *self, PyObject *args)
     return NULL;
 }
 
-PyDoc_STRVAR(functools_reduce_doc,
-"reduce(function, iterable[, initial], /) -> value\n\
-\n\
-Apply a function of two arguments cumulatively to the items of a sequence\n\
-or iterable, from left to right, so as to reduce the iterable to a single\n\
-value.  For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates\n\
-((((1+2)+3)+4)+5).  If initial is present, it is placed before the items\n\
-of the iterable in the calculation, and serves as a default when the\n\
-iterable is empty.");
-
 /* lru_cache object **********************************************************/
 
 /* There are four principal algorithmic differences from the pure python version:
@@ -1720,7 +1725,7 @@ PyDoc_STRVAR(_functools_doc,
 "Tools that operate on functions.");
 
 static PyMethodDef _functools_methods[] = {
-    {"reduce",          functools_reduce,     METH_VARARGS, functools_reduce_doc},
+    _FUNCTOOLS_REDUCE_METHODDEF
     _FUNCTOOLS_CMP_TO_KEY_METHODDEF
     {NULL,              NULL}           /* sentinel */
 };
```

You should run ``./python Tools/clinic/clinic.py Modules/_functoolsmodule.c`` to update autogenerated code.

</details>

Results:
| Benchmark            | ref     | patch                 | patch2                |
|----------------------|:-------:|:---------------------:|:---------------------:|
| reduce(f, lst)       | 2.18 us | 2.42 us: 1.11x slower | 2.11 us: 1.03x faster |
| reduce(f, lst, init) | 2.35 us | 2.64 us: 1.12x slower | 2.27 us: 1.04x faster |
| Geometric mean       | (ref)   | 1.12x slower          | 1.03x faster          |

Looks even slightly faster than in the main.

**Comment by sayandipdutta** [comment]
@skirpichev Is `initial=None` safe for backward compatibility? Does this mean `reduce(Callable[[None, T], None], Iterable[T], None)` will behave differently in 3.13 and 3.14?

File: Lib/functools.py
Comment (Eclips4) [inline Lib/functools.py]: This breaks backward compatibility, see https://github.com/python/cpython/pull/121677#discussion_r1676808256 for more information.

**Review by Eclips4 — COMMENTED** [review]
Thanks for the PR! I think that if we want to add keyword support for `functools.reduce`, it should be done for all parameters, not just the `initial`. If so, this would match the behavior of the pure Python version of functools.

**Comment by Eclips4** [comment]
> @skirpichev Is `initial=None` safe for backward compatibility? Does this mean `reduce(Callable[[None, T], None], Iterable[T], None)` will behave differently in 3.13 and 3.14?

No, it doesn't. Someone can use `None` as `initial` value.

File: Lib/functools.py
Comment (sayandipdutta) [reply Lib/functools.py]: Thanks for letting me know. I will revert this.

On the whole, I am +0 on enabling keyword support for all parameters in `reduce`. Personally I have not felt the need for other parameters barring `initial`. I can see the desire for parity with python version, but I don't see much readability gain from having `function` and `sequence` as keywords, as it is less confusing. Then again, I am generally against position only params for functions.

**Comment by skirpichev** [comment]
> it should be done for all parameters, not just the initial

That might slowdown the patch v2.

> No, it doesn't. Someone can use None as initial value.

That was a draft;)  I think we could use same trick as for the Python version.

BTW, it seems the [PEP 661](https://peps.python.org/pep-0661/) doesn't cover this at all.

File: Lib/functools.py
Comment (sayandipdutta) [reply Lib/functools.py]: I will change the actual signature. The docstring for python version of `reduce` shows the signature as positional only. Is it okay if I make the docstring signature to this?:
```diff
"""
-    reduce(function, iterable[, initial], /) -> value
+    reduce(function, iterable, /, initial=None) -> value
```
until there is agreement on whether to allow keywords for all params.

**Comment by sayandipdutta** [comment]
@skirpichev should the default be specified at all? I think `reduce(function, iterable, /, initial)` is closer representation of internal working than `reduce(function, iterable, /, initial=None)`. Or is there some sort of a convention?

File: Doc/library/functools.rst
Comment (sayandipdutta) [inline Doc/library/functools.rst:471]: I am unsure what to do about this. Should I mention `None`? Should I keep `/`? Any help is appreciated.

**Comment by skirpichev** [comment]
> I think reduce(function, iterable, /, initial) is closer representation

No.  Current code in the main more accurately can be described as function with multiple signatures.  Funny notation ``reduce(function, iterable[, initial], /)`` means it's possible to have *two* signature:
```
reduce(function, iterable, /)
reduce(function, iterable, initial, /)
```

The AC can't represent multiple signatures yet.  The only way - using the sentinel value ``_initial_missing``, like pure-Python version does.  See updated patch above.  You shouldn't use ``None`` as default value.

File: Doc/library/functools.rst
Comment (skirpichev) [reply Doc/library/functools.rst:471]: Nothing.

File: Lib/functools.py
Comment (nineteendo) [inline Lib/functools.py]: Maybe use ellipsis:

```suggestion
    reduce(function, iterable, /, initial=...) -> value
```

Comment (skirpichev) [reply Lib/functools.py]: See PEP 661:)

Comment (nineteendo) [reply Lib/functools.py]: But the sentinel is private and doesn't even exist in the C implementation. Ellipsis is frequently used for unspecified default values in typeshed. We could use multiple signatures though.

Comment (skirpichev) [reply Lib/functools.py]: > But the sentinel is private and doesn't even exist in the C implementation.

It's easy to add, see https://github.com/python/cpython/pull/125917#issuecomment-2434922118

> Ellipsis is frequently used for unspecified default values in typeshed.

@Eclips4?

> We could use multiple signatures though.

Yes, I think it's fine for the sphinx docs.  But help will looks like this (as for pure-Python version):
```pycon
>>> help(functools.reduce)
Help on built-in function reduce in module _functools:

reduce(function, iterable, /,
       initial=_functools._initial_missing)
    Apply a function of two arguments cumulatively to an iterable, from left to right.

    [...]
```

File: Lib/functools.py
Comment (skirpichev) [reply Lib/functools.py]: @sayandipdutta, you shouldn't use ``None`` as default value.

And, I think, there is no need to change sphinx docs further.  ``reduce(function, iterable, /[, initial]) -> value`` - just fine.

If you are going with AC version - see https://github.com/python/cpython/pull/125917#issuecomment-2434922118

Comment (sayandipdutta) [reply Lib/functools.py]: Ah, I just pushed. Okay, when everything else is sorted, I will fix this at the end.

File: Modules/_functoolsmodule.c
Comment (ZeroIntensity) [inline Modules/_functoolsmodule.c]: It would be nice to make this easier on the eyes.
```suggestion
calculates ((((1 + 2) + 3) + 4) + 5).
```

File: Modules/_functoolsmodule.c
Comment (ZeroIntensity) [inline Modules/_functoolsmodule.c]: I guess this is ok, but I would prefer to not use the private API where it doesn't help anything:
```suggestion
    if (PyModule_Add(module, "_initial_missing", PyObject_New(PyObject, &PyBaseObject_Type)) < 0) {
```

File: Doc/library/functools.rst
Comment (ZeroIntensity) [inline Doc/library/functools.rst]: We have this now :)
See [here](https://discuss.python.org/t/versionadded-next/65280)
```suggestion
   .. versionchanged:: next
```

File: Lib/functools.py
Comment (ZeroIntensity) [reply Lib/functools.py]: I don't think I've ever seen `=...` in the docs. Do we have precedent for that?

Comment (sayandipdutta) [reply Lib/functools.py]: It seems like the signature is giving `inspect` a hard time. But it is autogenerated by AC. Did I do something wrong?
```
reduce(function, iterable, /,
       initial=_functools._initial_missing)
```

Comment (Eclips4) [reply Lib/functools.py]: > But the sentinel is private and doesn't even exist in the C implementation. Ellipsis is frequently used for unspecified default values in typeshed. We could use multiple signatures though.


Multiple signatures for a docs sounds like a good solution.
Using `...` for default values is essentially the same as using `None`, and it's just wrong since users can pass `...` as the `initial` value.

File: Modules/_functoolsmodule.c
Comment (sayandipdutta) [reply Modules/_functoolsmodule.c]: I have made this change locally (didn't apply this because the changed line was too long), will push it with other changes. Thanks.

Comment (skirpichev) [reply Modules/_functoolsmodule.c]: Perhaps, a more pedantic version is:
```suggestion
    PyObject *missing = PyObject_New(PyObject, &PyBaseObject_Type);
    if (!missing || PyModule_Add(module, "_initial_missing", missing) < 0) {
        Py_XDECREF(missing);
```

Comment (ZeroIntensity) [reply Modules/_functoolsmodule.c]: Probably not--the benefit of `PyModule_Add` is that you don't need to do something like that, because it comes with a `NULL` check and steals the reference anyway. That looks like the kind of thing that would get used with `PyModule_AddObjectRef`

File: Lib/functools.py
Comment (skirpichev) [reply Lib/functools.py]: > I don't think I've ever seen =... in the docs. Do we have precedent for that?

Yeah, e.g. for the int.from_bytes, for example.

> it seems like the signature is giving inspect a hard time. But it is autogenerated by AC. Did I do something wrong?

First, note that reduce() has no correct signature in the current main.

Now AC adds one, but it can't be parsed by ``inspect._signature_fromstr()``: this helper has own opinion on what can be specified as a default value (e.g. it can't be a complex number).

File: Doc/library/functools.rst
Comment (skirpichev) [inline Doc/library/functools.rst]:
```suggestion
.. function:: reduce(function, iterable, /[, initial])
```

**Comment by nineteendo** [comment]
Do you update this test?

https://github.com/python/cpython/blob/ad6110a93ffa82cae71af6c78692de065d3871b5/Lib/test/test_inspect/test_inspect.py#L5699-L5701

**Comment by Eclips4** [comment]
> I've tried to implement @Eclips4 suggestion to make _all_ arguments positional-or-keyword, but this seems to be impossible with AC now. Looks as a bug.

What's the problem?

**Comment by sayandipdutta** [comment]
Yes, I just pushed the changes, i.e. made it unsupported instead of no_signature. I was trying look for docs regarding this, is there any? Like what makes a signature unsupported or no_signature?

File: Lib/functools.py
Comment (skirpichev) [inline Lib/functools.py]:
```suggestion
    reduce(function, iterable, /[, initial]) -> value
```

**Comment by skirpichev** [comment]
> Yes, I just pushed the changes, i.e. made it unsupported instead of no_signature.

Looks ok for me.

> I was trying to look for docs regarding this, is there any? Like what makes a signature unsupported or no_signature?

You can see _test_module_has_signatures() helper:
https://github.com/python/cpython/blob/2705ffa575ff12a27c9ff9d8be086df3071903a4/Lib/test/test_inspect/test_inspect.py#L5542

It *tests* signature as unsupported (e.g. because the ``__text_signature__`` added - can't be parsed by the inspect module).

> What's the problem?

@Eclips4, see https://github.com/python/cpython/issues/125954

Note also, that for pure-Python version in the master we have this declaration:
```py
def reduce(function, sequence, initial=_initial_missing):
```
while in the sphinx docs we have instead ``reduce(function, iterable[, initial], /)``.  What naming we can count as a part of the API, ``sequence`` or ``iterable``?

**Comment by JelleZijlstra** [comment]
For what it's worth I'm not convinced we need to make all parameters pos-and-keyword. I think there's a much stronger use case for passing `initial=` as a keyword argument, since it's an optional argument and even people who understand what `reduce` does in general might not immediately recognize it.

**Review by erlend-aasland — CHANGES_REQUESTED** [review]
> For what it's worth I'm not convinced we need to make all parameters pos-and-keyword. I think there's a much stronger use case for passing `initial=` as a keyword argument, since it's an optional argument and even people who understand what `reduce` does in general might not immediately recognize it.

+1

Moreover, I think that this change should be broken up in two steps:

1. Adapt `reduce()` to Argument Clinic (including benchmarks)
2. Allow `initial` to be a keyword argument

We know that when adapting functions and methods to Argument Clinic, it is easy to introduce subtle bugs, so I would really like this to be a separate change.

**Comment by Eclips4** [comment]
> > For what it's worth I'm not convinced we need to make all parameters pos-and-keyword. I think there's a much stronger use case for passing `initial=` as a keyword argument, since it's an optional argument and even people who understand what `reduce` does in general might not immediately recognize it.
> 
> +1
> 
> Moreover, I think that this change should be broken up in two steps:
> 
> 1. Adapt `reduce()` to Argument Clinic (including benchmarks)
> 2. Allow `initial` to be a keyword argument
> 
> We know that when adapting functions and methods to Argument Clinic, it is easy to introduce subtle bugs, so I would really like this to be a separate change.

I support the opinion that having all parameters as pos-and-keyword is not very useful, but if we only make `initial` pos-and keyword, then we need to decide what to do about Python implementation, which currently treats all parameters like pos-and-keyword arguments. Maybe we should start raising a warning if `function` and `sequence` arguments are passed as keyword arguments?

**Comment by erlend-aasland** [comment]
> [...] then we need to decide what to do about Python implementation [...]

Differences between `functools.py` and `_functoolsmodule.c` is out of scope for this issue/change.

**Comment by skirpichev** [comment]
@erlend-aasland as AC changes coming from my patch, probably 1) point - is my job.  I think that @sayandipdutta can keep this PR as-is for a while.  I'll make a separate PR with AC-related changes.

**Comment by skirpichev** [comment]
PR, that switch to AC: https://github.com/python/cpython/pull/125999

**Comment by sayandipdutta** [comment]
Thanks a lot @skirpichev! It seems all I have to do is wait for your PR to be merged and then merge main into my PR. Will do so.

**Comment by erlend-aasland** [comment]
Meta: I'll marked this as draft until Sergey's PR has landed.

**Comment by erlend-aasland** [comment]
> Meta: I'll marked this as draft until Sergey's PR has landed.

Sergey's Argument Clinic adaption landed just now. Please resolve conflicts, regenerate clinic, and mark the PR ready for review again.

File: Misc/NEWS.d/next/Library/2024-10-24-13-40-20.gh-issue-126916.MAgz6D.rst
Comment (erlend-aasland) [inline Misc/NEWS.d/next/Library/2024-10-24-13-40-20.gh-issue-126916.MAgz6D.rst]:
```suggestion
Allow the *initial* parameter of :func:`functools.reduce` to be passed as a keyword argument.
```

**Comment by erlend-aasland** [comment]
Can you post updated benchmarks vs. current `main`?

**Comment by sayandipdutta** [comment]
I have made the requested changes; please review again

**Comment by sayandipdutta** [comment]
Checked against debug build. Followed script from https://github.com/python/cpython/pull/125917#issuecomment-2434813259


|call| Main branch  | This branch |
|---| ------------- | ------------- |
|reduce(f, lst)| 3.59 us +- 0.13 us  | 3.51 us +- 0.14 us  |
|reduce(f, lst, initial)| 3.82 us +- 0.25 us  | 3.78 us +- 0.13 us  |

@erlend-aasland

**Review by erlend-aasland — APPROVED** [review]
Thanks!

**Comment by erlend-aasland** [comment]
BTW, you need a What's New entry for this.

**Comment by skirpichev** [comment]
@sayandipdutta, this now has merge conflicts.

**Review by skirpichev — APPROVED** [review]

**Comment by skirpichev** [comment]
I think that most people are ok with adding only one positional-or-keyword parameter (not 3).  Is there something to do, besides simple conflict resolution?

@Eclips4, are you ok with this (as your pr depends on the current one)?

**Review by Eclips4 — APPROVED** [review]
LGTM! Thank you!

**Comment by erlend-aasland** [comment]
Congrats with landing your PR, @sayandipdutta; good job, and thanks for your contribution! Thanks for reviewing, y'all!

**Comment by ZeroIntensity** [comment]
It looks like bedevere is stuck.

**Comment by sayandipdutta** [comment]
Thanks a lot everyone, for your help! Hope to be a regular contributor 🤞🏾

**Pull request merged**
